Created a guide for deploying DAGs on Airflow using Astronomer Core

jwitz opened the pull request:
This guide is one of the key components of our "Getting started" series for Astronomer Core. Essentially, we want to provide some basic guidelines for deploying DAGs at production scale. 

I wanted to give a sense of how to do this without being overly prescriptive. To do that, I gave instructions for the essential actions you need to complete, but leave it up to the reader to implement any automation tools they might want to use going forward.

kaxil commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+- Adding DAGs directly to a Docker image and rebuilding the image whenever DAGs are updated. Note that this method works only for users running Airflow on Kubernetes.
+
+
+## Automate DAG Deployment with a Cron Job
+
+We recommend using an automation tool to continuously pull new or updated DAGs to your Airflow machines. If you host your DAGs in a Git repository, you can use a cron job to pull new DAGs onto your machines on a regular basis. For each of your machines running Airflow:
+
+1. Open your DAGs directory and ensure that the folder is empty. If you followed the Installing at Production Scale guide, this folder would be `/usr/local/airflow`.
```
I feel like we should separate the deployment strategies where 

1) Deploying Airflow on Containers (managed via K8s and Helm)
2)Deploying Airflow on VMs (managed by Ansible)

kaxil replied:
Once we have that `deploy_dags` can be a sub-section of both of them

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+- Adding DAGs directly to a Docker image and rebuilding the image whenever DAGs are updated. Note that this method works only for users running Airflow on Kubernetes.
+
+
+## Automate DAG Deployment with a Cron Job
```
To @kaxil's point, this title might be better off a little bit more descriptive, so something like "Deploy Airflow on VMs". Is that the equivalent to this method?

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+
+## Automate DAG Deployment with a Cron Job
+
+We recommend using an automation tool to continuously pull new or updated DAGs to your Airflow machines. If you host your DAGs in a Git repository, you can use a cron job to pull new DAGs onto your machines on a regular basis. For each of your machines running Airflow:
```
Similar to my first comment, but it might be good to remind users what assumptions we're making when they get started on these steps. When they go to Step 1, what do we assume they have running already?

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+
+2. Clone your DAG repository into the folder using `$ git clone`.
+
+3. As your `airflow` system user, run `crontab -e`.
```
I'm personally having trouble with this command on a new MacOS ([this issue](https://superuser.com/questions/201172/mac-crontab-is-never-created/540693)) - did this work for you?

jwitz replied:
It did, you just need to press `i` to begin editing the document. Not sure whether this is a MacOS thing or a more general cron thing.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,91 @@
+    * */3 * * * cd /usr/local/airflow && git pull
+    ```
+
+Once you have this simple cron job in place, you can modify it to run on a per-merge basis using automation tools such as [Ansible](https://docs.ansible.com/ansible/latest/user_guide/index.html).
````
I'm not totally seeing a closed loop here - once the cron job is created, it will pull from your GitHub repository to `/usr/local/airflow`. And then what happens? Does a deploy to your Airflow Deployment happen every 3 minutes? Does your Webserver or Scheduler restart?

jwitz replied:
Shoot, I forgot that you have to restart the services in order to redeploy the DAGs. I updated the setup slightly to use an additional open source tool for checking if a directory has been modified. 

Are we assuming that anyone needing to deploy to VMs is using a Debian-like OS as described in the guide? Some of this open source setup can get dicey when it needs to be set up for both Mac and Linux.

paolaperaza replied:
That's a good question - I'm actually not sure, but I'd imagine not a small percentage of users are on Mac. Do we have an idea of what these steps would be for other OS?

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+
+If you run Airflow in Docker, you can deploy DAGs via the [Astronomer Core Helm chart](https://github.com/astronomer/airflow-chart). This is also known as "baking in" DAGs because you're adding DAGs to the image itself.  To use this method, you'll need:
+
+- A Dockerfile that pulls Astronomer Core's [Docker image](https://hub.docker.com/r/astronomerinc/ap-airflow).
+- A Kubernetes deployment.
+- The kubectl command line tool.
+- Helm.
+- A Docker registry.
```
```suggestion
- A Dockerfile that pulls Astronomer Core's [Docker image](https://hub.docker.com/r/astronomerinc/ap-airflow)
- A Kubernetes Cluster
- The [Kubernetes CLI (kubectl)](https://kubernetes.io/docs/tasks/tools/)
- Helm
- A Docker registry
```

No period on these, since they're not full sentences?

jwitz replied:
Some folks still use periods when the bullet completes an [introductory stem](https://lawprofessors.typepad.com/legal_skills/2012/01/how-to-punctuate-bullet-points-.html#:~:text=Punctuating%20Bullet%20Points&text=Use%20a%20period%20(full%20stop,do%20not%20complete%20the%20stem), but I agree that it looks better this way and probably represents more common usage.

paolaperaza replied:
Ah, got it - I'm cool either way, I just had the impression we had made that call earlier to only do periods with full sentences and not with a "list" of objects. As long as we're consistent, I'm good.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,91 @@
+4. Build the Docker image using the following command:
+
+    ```sh
+    docker build -t <filepath>/ap-airflow:<your-image-tag>
````
What's the `filepath` here? Is that the user's docker registry, or just `quay.io/astronomer`?

paolaperaza replied:
We also have `<your-image-tag>` here, but `latest-onbuild` in the step above. Should we have a brief note that the tag here should reflect the version of Airflow you'd like to run (worded more precisely), and then link to doc?

jwitz replied:
`filepath` is maybe confusing, I just meant `etc`/ whatever it takes to get to `ap-airflow`.

This brings up a question I've had when doing the quickstart: if we assume that someone will use our Docker image in production, do we need to tell them to update `latest-onbuild` to a specific release, lest they continuously upgrade without realizing it?

paolaperaza replied:
I'd say assume `latest-onbuild` for now, and somewhere we'll have a note that directs people to the "Upgrade Apache Airflow" doc for more info. WDYT?

paolaperaza replied:
Still not understanding what `file-path` is though - I'd think our registry would have to be specified there since that's where it's pulling from?

jwitz replied:
Oops, those should also be `etc` to denote that the image you're building is coming from a local Dockerfile. The quay.io path is specified only in the Dockerfile itself.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,91 @@
+    --set images.airflow.tag=<your-image-tag>
+    ```
+
+With this method, you no longer need to store your DAG folder on each individual machine running Airflow. While baking DAGs into an image is a great way to keep all of the code for your project running in one place, it can be labor-intensive to rebuild your image each time you update a DAG.
````
This is actually similar to how the Astronomer Platform handles deploys. Similar to my question above - does this method restart the Webserver/Scheduler?

jwitz replied:
Nope, don't think so. Added another step for this.

paolaperaza reviewed (commented):
@jwitz Thanks for getting this started! Jotted down some initial, mostly educational questions as we get in the weeds of the process. If you know the answers to any of these, feel free to add info directly. If not, happy for us to do the research together!

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+
+## Overview
+
+When running Airflow at production scale, every machine running Airflow needs a copy of your DAG files. In addition, these DAG files all need to appear in the same directory. This guide will cover two popular methods for deploying DAGs across your machines:
```
Tried to make this a bit clearer. I suppose it's less the machines running Airflow, as Docker could also be run on a machine that wasn't installed as part of the production installation. 

I'm hoping that the sidebar of the doc architecture on our website will suggest that this setup is completed after doing a Docker or Python installation.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,91 @@
+
+1. Open your DAGs directory and ensure that the folder is empty. If you followed the Installing at Production Scale guide, this folder would be `/usr/local/airflow`.
+
+2. Clone your DAG repository into the folder using `$ git clone`.
```
Reworded this slightly.

jwitz commented:
@paolaperaza This should be ready for another look! Changed the setup for virtual machine deployments and updated our wording around "Production" as we discussed earlier (with the exception of the to-be-published installation guide links). Any other non-straightforward changes should be marked with a comment.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+
+## Overview
+
+Whether you're running Airflow on virtual machines or in Docker, every node in your Airflow cluster needs a copy of your DAG files. In addition, these DAG files all need to appear in the same directory. This guide will cover two popular methods for deploying DAGs across your machines:
```
Apache Airflow docs refer to multiple machine setups as "multi node clusters".. is that something we want to use in our own docs?

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+
+Whether you're running Airflow on virtual machines or in Docker, every node in your Airflow cluster needs a copy of your DAG files. In addition, these DAG files all need to appear in the same directory. This guide will cover two popular methods for deploying DAGs across your machines:
+
+- Using a cron job to regularly pull DAGs into your local directories from a central Git repository.
+- Adding DAGs directly to a Docker image and rebuilding the image whenever DAGs are updated. To implement this method, Kubernetes is required.
```
```suggestion
- Deploy Airflow on Virtual Machines by using a cron job that pulls DAGs into your local directories from a central Git repository.
- Deploy Airflow on Containers by building DAGs into a Docker image and rebuilding the image whenever DAGs are updated. To implement this method, Kubernetes and Helm are required.
```

@jwitz Made this suggestion as I read this a second time, and I think it more closely aligns with a comment Kaxil made earlier. We could also apply these buckets to the H2's below, so:

```
## Deploy Airflow DAGs on VMs via Cron Job
## Deploy Airflow DAGs on Containers via Helm Chart
```

jwitz replied:
Sounds good, only change I made here was not capitalizing virtual machine

paolaperaza replied:
Cool

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+
+## Deploy DAGs to Multiple Machines via Cron Job
+
+If you installed Airflow on a virtual machine, you can use a simple cron job, as well as open source file detection tools like `incron`, to pull new DAGs onto your Airflow machines on a regular basis and restart Airflow when necessary. This setup assumes that you:
```
```suggestion
If you installed Airflow on a virtual machine, you can use a simple cron job, as well as open source file detection tools like `incron`, to regularly pull new DAGs onto your Airflow machines and restart Airflow when those DAG files are updated. This setup assumes that you:
```

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+If you installed Airflow on a virtual machine, you can use a simple cron job, as well as open source file detection tools like `incron`, to pull new DAGs onto your Airflow machines on a regular basis and restart Airflow when necessary. This setup assumes that you:
+
+- Installed Airflow via Python wheel according to [Installing at Production Scale].
+- Host your DAGs in a Git repository.
```
```suggestion
- Host your DAGs in a GitHub repository.
```

jwitz replied:
It doesn't have to be GitHub. There are other repos and structures, like GitLab, that still use git commands as described here. GitHub is the host, but Git is the protocol. Someone might set up an entirely local Git repo for security reasons: https://git-scm.com/book/en/v2/Git-Basics-Getting-a-Git-Repository

paolaperaza replied:
Ah ah ah - you're totally right. In that case, might it be better to abstract "git" entirely and say your "version control tool"? Or assume Git and either GitLab or GitHub? Good call either way - will leave up to you

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+
+1. Open your DAGs directory and ensure that the folder is empty. If you followed the Installing at Production Scale guide, this folder would be `/usr/local/airflow`.
+
+2. Clone the Git repository holding your DAGs into the folder using `git clone`.
```
```suggestion
2. Clone the GitHub repository holding DAGs into your `/usr/local/airflow` folder using `git clone`.
```

jwitz replied:
See comment above on Git vs GitLab. 

If users can name `/usr/local/airflow` something else, should we keep the filename generic in setups? I only included it above for reference if they were coming from the installation guide. I guess I'm not sure how common it would be for someone to name it something else, or how strongly we should suggest not doing that.

paolaperaza replied:
Fair question - I'm not sure to be honest, but I can imagine at least _some_ people doing so. Maybe keep the file path out, like you said, but specify what "the folder" is supposed to be. So maybe:

> 2. Clone the Git repository holding your DAGs into your Airflow project directory by using `git clone`.

jwitz replied:
But it's not the project directory they're cloning into, but the DAG directory itself right? We might be overthinking this too much, but I edited the first step so that it's guaranteed that readers are in the right directory when they run `git clone`.

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+
+3. As your `airflow` system user, run `crontab -e` to create a new cron job. You can specify an editor using the `EDITOR` environment variable (e.g. `env EDITOR=atom crontab -e`).
+
+4. Create a cron job that continuously pulls from the DAG repository. For example, if you want to pull to the `/usr/local/airflow` folder every 3 minutes, your cron job would look something like this:
```
```suggestion
4. Create a cron job that regularly pulls from the GitHub repository in which your DAG files are hostd. For example, if you want to pull from GitHub to your `/usr/local/airflow` directory every 3 minutes, your cron job would look something like this:
```

jwitz replied:
See comment above on Git. To keep the step simple, and since we mention the use of the Git folder above, I removed the "in which your DAG files are hosted" part of this.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,109 @@
+    /usr/local/airflow IN_MODIFY systemctl restart airflow
+    ```
+
+    This cron job first pulls from a Git repository, then checks to see if any files in the repository have been modified since the last pull. If a file was modified, the cron job restarts Airflow so that the updated DAGs are successfully deployed.
````
```suggestion
    This cron job first pulls from a Git repository, then checks to see if any files in the repository have been modified since the last pull. If a file was modified, the cron job restarts Airflow and triggers a deploy to your Airflow environment with your updated DAGs.
```

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,109 @@
+
+Once you have this simple cron job saved, you can scale it and use tools such as [Ansible](https://docs.ansible.com/ansible/latest/user_guide/index.html) to integrate into a larger CI/CD workflow. If you are running MacOS, you can complete a similar setup using [kqwait](https://github.com/sschober/kqwait).
+
+## Deploy DAGs on Kubernetes via Helm Chart
```
```suggestion
## Deploy DAGs on Kubernetes via Helm
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,109 @@
+3. Edit your Dockerfile with the following minimum lines:
+
+    ```
+    # Use a different tag to run a different version of Airflow
````
```suggestion
    # Use a different tag to specify a supported Apache Airflow version
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,109 @@
+    docker build -t <filepath>/ap-airflow:<your-image-tag>
+    ```
+
+    As a best practice, the image tag you specify here should indicate which version of Airflow you're using.
````
```suggestion
    As a best practice, the image tag you specify here should indicate which version of Airflow you're using. To run Apache Airflow 2.0.1, for example, this command would be:
      ```
     docker build -t <filepath>/ap-airflow:2.0.1
      ```
```

paolaperaza reviewed (commented):
A few more general comments for ya @jwitz

jwitz commented:
@paolaperaza some changes, replies, and additional comments in this latest commit!

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,113 @@
+
+## Deploy DAGs on Kubernetes via Helm
+
+If you run Airflow in Docker, you can deploy DAGs via the [Astronomer Core Helm chart](https://github.com/astronomer/airflow-chart). This is also known as "baking in" DAGs because you're adding DAGs to the image itself.  To use this method, you'll need:
```
Might it be good to add more context (either here or even just under ## Overview) around the pros, cons of these two deploy methods and why we might recommend one over the other? For example, it seems like we can _generally_ recommend deploying DAGs on Kubernetes via Helm for production workloads.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+2. Install the Helm chart using the following command:
+
+    ```sh
+    helm install --name <your-release-name>
````
Question for my own understanding - what does `--name` here stand for?

jwitz replied:
It's just a flag for announcing the release name, I think it's required in later versions of Helm.

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,113 @@
+
+## Deploy DAGs on Kubernetes via Helm
+
+If you run Airflow in Docker, you can deploy DAGs via the [Astronomer Core Helm chart](https://github.com/astronomer/airflow-chart). This is also known as "baking in" DAGs because you're adding DAGs to the image itself.  To use this method, you'll need:
```
```suggestion
If you run Airflow in Docker, the easiest transition to a production-grade workflow is often to deploy DAGs onto a Kubernetes environment via the [Astronomer Helm chart](https://github.com/astronomer/airflow-chart), or another community-provided Helm chart. Here, you would build both your DAGs and the rest of your Airflow project directory into a single Docker image that would then be pushed to your Docker registry and passed to each of your core Airflow components for execution.

To use this method, you'll need:
```

paolaperaza replied:
Soft suggestion for potential additional context

jwitz replied:
Your suggestion here quietly includes a can-of-worms question: is this doc just for people running AC, or for all Airflow Users? If it's just AC, I think it's fine for us to suggest using our own Helm chart and refrain from mentioning other options, as that will hopefully one day include the best support and compatibility with AC. 

Reworded slightly under the assumption that we want to promote AC's own open source solutions for running Airflow, lmk what you think.

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,113 @@
+    # Use a different tag to specify a supported Apache Airflow version
+    FROM quay.io/astronomer/ap-airflow:latest-onbuild
+
+    COPY <your-dag-directory> ~astro/airflow-venv
```
Question for my own edification - Does this only copy your DAG directory? What if you wanna build other stuff into your image?

jwitz replied:
You would, but it would only be other dependencies, which you could also just build into your machines running Airflow. Some Dockerfiles copy `requirements.txt` in (like the one generated by `astro-dev-init`) but you could also install those on machines. Maybe we could edit this section to just tell users to add the `COPY` line for DAGs, since they presumably have the rest of the Dockerfile configured as needed?

jwitz replied:
I think so just based on the fact that we do the same thing with our `astro dev init` generated image. I think as a general practice it's good to have everything you need ready to go in your container. 

I think we'd be better off giving this advice in the `install dependencies` guide, since that will have more in-depth information on each of these files. I could link out to that page from the bottom of the doc in a `Next Steps` section, but I don't think we should distract from the baking in setup by including it within a step.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+    docker build -t etc/ap-airflow:<your-image-tag>
+    ```
+
+    As a best practice, the image tag you specify here should indicate which version of Airflow you're using. To run Apache Airflow 2.0.1, for example, this command would be:
````
```suggestion
    `docker-airflow:latest-onbuild` will pull the latest version of Apache Airflow supported by Astronomer Core. To specify a particular version of Airflow you'd like to run, replace `your-image-tag`. To run Apache Airflow 2.0.1, for example, this command would be:
```

`docker-airflow` will def be the name of our docker repos, so let's just replace for now if you agree?

jwitz replied:
Sure, but this isn't where you actually pull the image. This is where you write your own custom image tag that, as a best practice, should reflect what makes this image unique. That could be a version number or any number of words that indicate the customization you applied. 

Now that I took the other FROM line out for the Dockerfile snippet, I think we should cut this bit entirely. It's probably better to mention this briefly in the Dependencies guide and keep these steps focused on DAGs.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+4. Build the Docker image using the following command:
+
+    ```sh
+    docker build -t etc/ap-airflow:<your-image-tag>
````
```suggestion
    docker build -t etc/docker-airflow:<your-image-tag>
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+
+    ```
+    # Use a different tag to specify a supported Apache Airflow version
+    FROM quay.io/astronomer/ap-airflow:latest-onbuild
````
```suggestion
    FROM quay.io/astronomer/docker-airflow:latest-onbuild
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+    As a best practice, the image tag you specify here should indicate which version of Airflow you're using. To run Apache Airflow 2.0.1, for example, this command would be:
+
+    ```sh
+    docker build -t etc/ap-airflow:2.0.1
````
```suggestion
    docker build -t etc/docker-airflow:2.0.1
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+5. Push the Dockerfile to your registry with the following command:
+
+    ```sh
+    docker push etc/ap-airflow:<your-image-tag>
````
```suggestion
    docker push etc/docker-airflow:<your-image-tag>
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+
+    ```sh
+    helm upgrade <your-release-name> . \
+    --set images.airflow.repository=etc/ap-airflow \
````
```suggestion
    --set images.airflow.repository=etc/docker-airflow \
```

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,113 @@
+    --set images.airflow.tag=<your-image-tag>
+    ```
+
+7. `cd` to your Airflow project folder and run `astro dev stop && astro dev start` to restart your Airflow services.
````
These commands only work locally, right? Not if you're managing an Airflow Deployment on some K8s cluster (but yes if they're just using the Astro CLI locally)

jwitz replied:
Ugh yeah this varies depending on setup.. restarting with docker compose is a separate command, while k8s doesn't need a restart because the new pods will just begin spinning up with the new image (might be wrong about that). Changed this to be much more generic.

paolaperaza replied:
^Realized in my latest comment I felt that we needed more specific instructions here. There's gotta be a "right answer' on what users need to do here to regularly push DAGs?

paolaperaza reviewed (commented):
Thanks, @jwitz - important back and forth we got going on here. Sent over a couple more questions/suggestions!

jwitz commented:
These are great questions @paolaperaza and while we don't know all the answers yet, I did my best to answer them in this latest commit.

I noticed another instance of ap-airflow in the Dockerhub link in prerequisites. Assuming that would be updated, I changed it even though it 404's at the moment. Problem is, our link checker doesn't seem to catch that 🤐  maybe something to look into.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,106 @@
+
+For this setup, you'll need:
+
+- A Dockerfile that pulls Astronomer Core's [Docker image](https://hub.docker.com/r/astronomerinc/docker-airflow)
```
Note for self: fix link

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,106 @@
+    helm repo add astronomer https://helm.astronomer.io
+    ```
+
+2. Install the Helm chart using the following command:
````
We may want to specify which Helm chart we're installing here and link people to our source code since there are a few Helm charts one could in theory use for this function.

In this case, we're sending them to the Astronomer Helm Chart: https://github.com/astronomer/airflow-chart

We could add, "This is Astronomer's Helm Chart, which is built off of the Apache Airflow Project's community chart and allows for more extensibility." (Not quite true today smh but it will be - either way, at least have half a sentence here I'm thinking)

Should we give them the option of installing another Helm Chart? (i.e. [Bitnamis](https://github.com/bitnami/charts/tree/master/bitnami/airflow), or [Airflow's](https://github.com/apache/airflow/tree/master/chart)?)

jwitz replied:
We have that link in the intro to this setup, and the link from step 1 also downloads the Helm chart. Might be redundant to also include it in the steps. Added your point about extensibility to that intro part.

Cant find my previous comment on this, but I think this introduces a higher level question on how much we want to promote staying within the open source Astro environment. Imo, if we can't provide meaningful advice on when to use one chart over another, then we should just focus on our own chart.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,106 @@
+    COPY <your-dag-directory> ~astro/airflow-venv
+    ```
+
+4. Build the Docker image using the following command:
````
I think similar comment here as the Helm chart - link out to source code for this docker image and describe Astronomer Core very briefly? https://github.com/astronomer/docker-airflow

jwitz replied:
I can add this link to the prerequisites, but again I'm not keen on adding a lot of background info within the steps themselves.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,106 @@
+    docker push etc/docker-airflow:<your-image-tag>
+    ```
+
+6. Upgrade your image using the following Helm command:
````
Looking at Helm docs, and I don't know that "image" is the right terminology here - maybe:

"Apply your changes by upgrading your release with the following Helm command:"

https://helm.sh/docs/intro/using_helm/

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,106 @@
+5. Push the Dockerfile to your registry with the following command:
+
+    ```sh
+    docker push etc/docker-airflow:<your-image-tag>
````
I may have asked this question before, but this `etc` is still throwing me off - should we tell people to replace `etc` this with the name of their registry? Or is it not ours?

jwitz replied:
Yeah it's their own local filepath. We do this in other guides too, I honestly don't know what the best placeholder value for indeterminate filepaths are. <filepath> felt weird as well.

How do you feel about `path/to/file`? I like that because it's obvious that it's not a real pathfile, but also maintains the correct syntax

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,106 @@
+
+7. Restart your Airflow services.
+
+With this method, you no longer need to store your DAG folder on each individual machine running Airflow. While baking DAGs into an image is a great way to keep all of the code for your project running in one place, it can be labor-intensive to rebuild your image each time you update a DAG.
```
The point being made is a little confusing here, since it ends on a "negative" note. How about:

> When you bake your DAGs into a single Docker image and deploy that image to a Kubernetes environment using Helm, you don't need to install dependencies and store your DAGs on each individual machine running Airflow in the way you would on a VM. With that said, needing to re-build your image every time you update a DAG file can be labor-intensive. As a next step, we recommend hosting your DAG folder in a Git repository and implementing a CI/CD tool to automatically rebuild your image whenever code is pushed or merge to the repository.

paolaperaza commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,106 @@
+    --set images.airflow.tag=<your-image-tag>
+    ```
+
+7. Restart your Airflow services.
````
Do we want to include instructions here about how to restart Airflow? It seems slightly unclear what users should do when they just want to update a DAG file once they've installed AC and cloned our Helm repo for the first time.

Do they have to go through these 1-7 steps again, or just the last few?

jwitz replied:
First answer: given that they could be using any cloud service to manage these services, I don't think we can provide explicit commands the way we do for the Python wheel setup. 

Second answer: That's a good point, it's unclear from these steps. I split them into two separate lists: one for initial setup, and another for actually baking in the DAGs.

paolaperaza reviewed (commented):
A few more comments here, @jwitz :)

jwitz commented:
Great questions @paolaperaza helped me get a lot more of this doc figured out. Lingering questions about referencing our own Helm chart aside, I think all of your comments have been addressed!

paolaperaza reviewed (commented):
A few more comments here, @jwitz - but we're close!

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+- Host your DAGs in a Git repository.
+- Have a system user named `airflow`.
+- Have [incron](https://github.com/sschober/kqwait) installed on all machines running Airflow.
+- Are using a Debian-like OS.
```
Alsoo, is there an actual technical definition for "Debian-like OS"

jwitz replied:
we could probably change this to "debian-based", which is probably the most common way to reference that.  I think there are different Linux flavors, some of which are based on Debian. 

Should we mention `Systemctl` even though that's how we install Airflow in our guide? I guess we mention that you need a system user named `airflow` in the prereqs, which is also in the install guide.. I'm not sure how much we want to assume a persistent installation across our docs.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+
+## Overview
+
+Whether you're running Airflow on virtual machines or in Docker, every node in your Airflow cluster needs a copy of your DAG files. In addition, these DAG files all need to appear in the same directory. This guide will cover two popular methods for deploying DAGs across your machines:
```
Agree with all of these calls!

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+- Deploy Airflow on Virtual Machines by using a cron job that pulls DAGs into your local directories from a central Git repository.
+- Deploy Airflow on Containers by building DAGs into a Docker image and rebuilding the image whenever DAGs are updated. To implement this method, Kubernetes and Helm are required.
+
+## Deploy DAGs to Multiple Machines via Cron Job
```
There's other ways you could do it, but we don't recommend them here. 

Side note: If we publish an Ansible role, we definitely want to build this functionality into that.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+
+## Deploy DAGs to Multiple Machines via Cron Job
+
+If you installed Airflow on a virtual machine, you can use a simple cron job, as well as open source file detection tools like `incron`, to regularly pull new DAGs onto your Airflow machines and restart Airflow when those DAG files are updated. This setup assumes that you:
```
Cut this down a bit.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+
+1. `cd` to your local DAGs directory and ensure that the folder is empty. If you followed the Installing at Production Scale guide, this folder would be `/usr/local/airflow`.
+
+2. Clone your Git repository for DAGs into the folder using `git clone`.
```
Made this active voice

jwitz commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,112 @@
+    /usr/local/airflow IN_MODIFY systemctl restart airflow
+    ```
+
+    If a file was modified from the pull, the cron job restarts Airflow and triggers a deploy to your Airflow environment with any updated DAGs.
````
Coupled with the actual command's text, I think we can keep this description simple.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+5. Run `systemctl start incron.service` to begin monitoring the DAG directory for changes.
+
+Once you have this simple cron job saved, you can scale it and use tools such as [Ansible](https://docs.ansible.com/ansible/latest/user_guide/index.html) to integrate into a larger CI/CD workflow. If you are running MacOS, you can complete a similar setup using [kqwait](https://github.com/sschober/kqwait).
+
```
We say the cron job is an example above, so I think this note is self-evident. I included a link to the crontab guru site for when we first mention creating a cron job.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+
+## Deploy DAGs on Kubernetes via Helm
+
+If you run Airflow in Docker, the most efficient transition to a production-scale workflow is to deploy DAGs onto a Kubernetes environment via the [Astronomer Helm chart](https://github.com/astronomer/airflow-chart), which is built off of the Apache Airflow Project's community chart to provide more extensibility.
```
I agree that `with` is safest, but I don't think we need to formalize it? I think any usage is correct, as in people will understand what is meant regardless of what preposition goes here.

jwitz commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,112 @@
+
+If you run Airflow in Docker, the most efficient transition to a production-scale workflow is to deploy DAGs onto a Kubernetes environment via the [Astronomer Helm chart](https://github.com/astronomer/airflow-chart), which is built off of the Apache Airflow Project's community chart to provide more extensibility.
+
+With this setup, you build a custom Docker image that contains your DAGs, also known as "baking" DAGs into the image. This custom image is then pushed to your Docker registry and passed to each of your core Airflow components for execution.
```
This is getting to be a pretty long explanation for something pretty simple. I tried rewriting it to be shorter and simpler.

jwitz commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,112 @@
+    COPY <your-dag-directory> ~astro/airflow-venv
+    ```
+
+Then, for every time that you update a DAG:
````
Do we need to say "want to deploy your changes"? I assume that's what they want to do if they're updating a DAG and reading this guide.

jwitz commented on ac/next/04_deploy-dags.md:
````diff
@@ -0,0 +1,112 @@
+    helm install --name <your-release-name>
+    ```
+
+    To find your release name for Airflow, you can run:
````
Yeah, although we describe this same thing on the platform as a release name. Going forward, we should probably say namespace just to be consistent with k8s (defos putting that in the style guide).

paolaperaza commented on ac/next/04_deploy-dags.md:
```diff
@@ -0,0 +1,111 @@
+
+If you installed Airflow on a VM, you can use a cron job and an open source file detection tool like `incron` to regularly deploy new DAGs to your Airflow instance. This setup assumes that you:
+
+- Manage your Airflow services using Systemd as described in [Install Apache Airflow on a Virtual Machine].
```
```suggestion
- Manage your Airflow services using [systemd](https://systemd.io/) as described in [Install Apache Airflow on a Virtual Machine].
```

Sorry I'm the worst but it does look systemd refers to itself in lower-case

paolaperaza reviewed (approved):
1 more minor comment, but I think we're ready to rock here @jwitz ! Merge whenever you're ready and close out that story in Clubhouse 😊
